virttest/migration.py

```diff
@@ -270,27 +282,159 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
         :param options: migration options
         :param thread_timeout: time out seconds for the migration thread running
         :param ignore_status: determine if an exception is raised for errors
-        :param func: the function executed during migration thread is running
+        :param func: the function or list of functions executed during migration
+                     thread is running
         :param args: dictionary used by func,
                      'func_param' is mandatory if no real func_param, none is
                      requested.
                      'shell' is optional, where shell=True(bool) can be used
                      for process.run
 
         """
+
+        def _run_collect_event_cmd():
+            """
+            To execute virsh event command to collect the domain events
+
+            :return: VirshSession to retrieve the events
+            """
+            cmd = "event --loop --all"
+            virsh_event_session = virsh.VirshSession(virsh_exec=virsh.VIRSH_EXEC,
+                                                     auto_close=True,
+                                                     uri=srcuri)
+            virsh_event_session.sendline(cmd)
+            logging.debug("Begin to collect domain events...")
+            return virsh_event_session
+
+        def _need_collect_events(funcs_to_run):
+            """
+            Check if there is a need to run a command to collect domain events
+
+            :param funcs_to_run: the functions to be run. It can be a list
+                                 or a single function
+                For example,
+                funcs_to_run = [{"func": <function check_established at 0x7f5833687510>,
+                                 "begin_event": "iteration: '1'",
+                                 "end_event": "Suspended Migrated",
+                                 "before_pause": "yes"},
+                                {"func": <function domjobabort at 0x7f5835cd08c8>,
+                                 "before_pause": "no"}
+                               ]
+            :return: boolean, True to collect events, otherwise False
+            """
+            if not funcs_to_run:
+                return False
+            if isinstance(funcs_to_run, list):
+                for one_func in funcs_to_run:
+                    if isinstance(one_func, dict):
+                        begin_event = one_func.get('begin_event')
+                        end_event = one_func.get('end_event')
```

Comment:
Can we set more specific argument names? for example, func_before_event or func_after_event...

Reply:
updated